### DISCUSSION TITLE
ModuleNotFoundError: No module named 'igor'

### DISCUSSION THREAD

**Discussion by ns-rse** [AFM-SPM/TopoStats#475] [Q&A]
[Original Question](https://github.com/AFM-SPM/TopoStats/discussions/39#discussioncomment-5027875)

> Hi,
> I'm Dake Mao and I'm working on measuring DNA cyclic/acyclic fibers with Topostats. I previously > used Topostats in Docker and it helped a lot. However, I met some questions in running Topostats > by PyPI.
> After I create the virtual environment by Miniconda, I installed the Topostats by “pip install > topostats”. Then I run the "run_topostats". The commander lines show the following results:
> 
> D:\Program Files\python\Lib\site-packages\topostats>run_topostats
> Traceback (most recent call last):
> File "d:\program files\python\lib\runpy.py", line 197, in _run_module_as_main
> return _run_code(code, main_globals, None,
> File "d:\program files\python\lib\runpy.py", line 87, in run_code
> exec(code, run_globals)
> File "D:\Program Files\python\Scripts\run_topostats.exe_main.py", line 4, in
> File "d:\program files\python\lib\site-packages\topostats\run_topostats.py", line 21, in
> from topostats.io import read_yaml, write_yaml, LoadScan
> File "d:\program files\python\lib\site-packages\topostats\io.py", line 9, in
> from igor import binarywave # this is already imported with afmformats but is being specialised > for our use
> ModuleNotFoundError: No module named 'igor'
>
> I'm wondering if you could help me to solve this error. Thanks.
> 
> Best wishes,
> Dake Mao

**Comment by ns-rse** [comment]
Hi @DakeMaoDNA,

Thanks for getting in touch.

It looks like this is happening because we omitted to include the Python package [`igor`](http://blog.tremily.us/posts/igor/) as a dependency for TopoStats. This has been fixed in the development version.

There are two possible options...

## Install Igor

You could install Igor under your virtual environment with the following command and `run_topostats` should then work.

```bash
conda activate <name_of_your_virtual_env>
pip install igor
run_topostats -c <your_config>.yaml
```

## Try the development version of TopoStats

Alternatively you could try the development version of TopoStats and install from GitHub. There are instructions on the [TopoStats : Installation](https://afm-spm.github.io/TopoStats/installation.html#cloning-from-github).

If you try this option and have any problems please do not hesitate to reply in this thread.

Please let us know how you get on and if you've any more questions or problems (if its a new problem please start a [new discussion in Q&A](https://github.com/AFM-SPM/TopoStats/discussions/new?category=q-a)).

Neil

**Reply by DakeMaoDNA** [reply]
Hi Neil,

Thanks for the solution. I have solved the Igor problem by installing the Igor as in your reply. However, I met another question with Numpy in complex numbers. I tried to install different versions of Numpy but it didn't help. Here is the result:

(topostats) D:\>run_topostats
Traceback (most recent call last):
  File "d:\program files\python\lib\runpy.py", line 197, in _run_module_as_main
    return _run_code(code, main_globals, None,
  File "d:\program files\python\lib\runpy.py", line 87, in _run_code
    exec(code, run_globals)
  File "D:\Program Files\python\Scripts\run_topostats.exe\__main__.py", line 4, in <module>
  File "d:\program files\python\lib\site-packages\topostats\run_topostats.py", line 21, in <module>
    from topostats.io import read_yaml, write_yaml, LoadScan
  File "d:\program files\python\lib\site-packages\topostats\io.py", line 9, in <module>
    from igor import binarywave  # this is already imported with afmformats but is being specialised for our use
  File "d:\program files\python\lib\site-packages\igor\binarywave.py", line 110, in <module>
    1:_numpy.complex, # NT_CMPLX, makes number complex.
  File "d:\program files\python\lib\site-packages\numpy\__init__.py", line 305, in __getattr__
    raise AttributeError(__former_attrs__[attr])
AttributeError: module 'numpy' has no attribute 'complex'.
`np.complex` was a deprecated alias for the builtin `complex`. To avoid this error in existing code, use `complex` by itself. Doing this will not modify any behavior and is safe. If you specifically wanted the numpy scalar type, use `np.complex128` here.
The aliases was originally deprecated in NumPy 1.20; for more details and guidance see the original release note at:
    https://numpy.org/devdocs/release/1.20.0-notes.html#deprecations

Best wishes,
Dake Mao

**Reply by ns-rse** [reply]
Hi @DakeMaoDNA 

We have encountered this ourselves and it arises with newer versions of Numpy. In the development version we pegged the Numpy version on which TopoStats depends and the error does not occur.

For the time being you will have to try installing the development version from GitHub (solution 2 above).

We are finishing off a few pieces of work at the moment though and expect to make a new release to PyPi that includes these and a number of other fixes as well as some new functionality in the not too distant future.

Let me know if you have any trouble installing from GitHub, you may want to `pip uninstall topostats` from your Conda environment first though so there is no conflict.

Regards

@ns-rse / Neil

**Reply by DakeMaoDNA** [reply]
Hi Neil,

Thanks! Now the development version perfectly works on my computer and datasets.

Best wishes,
Dake Mao

**Reply by ns-rse** [reply]
That is great to hear @DakeMaoDNA :+1:
